src/cryocanvas/app.py

```diff
@@ -151,143 +157,167 @@ def get_prediction_layer(self):
-            training_features = compute_features(
-                mask_idx, use_skimage_features, use_tomotwin_features
             )
-            training_labels = np.squeeze(active_labels)
-        elif data_choice == "Whole Image":
-            if use_skimage_features:
-                training_features = np.array(self.feature_data_skimage)
-            else:
-                training_features = np.array(self.feature_data_tomotwin)
-            training_labels = np.array(self.painting_data)
+        if features:
+            return np.concatenate(features, axis=1)
         else:
-            raise ValueError(f"Invalid data choice: {data_choice}")
-
-        if (training_labels is None) or np.any(training_labels.shape == 0):
-            self.logger.info("No training data yet. Skipping model update")
-        elif live_fit:
-            # Retrain model
-            self.logger.info(
-                f"training model with labels {training_labels.shape} features {training_features.shape} unique labels {np.unique(training_labels[:])}"
-            )
-            self.model = self.update_model(
-                training_labels, training_features, model_type
-            )
+            raise ValueError("No features selected for computation.")
 
+    @ensure_main_thread
+    def on_model_fit(self, task: Future, *, live_prediction: bool, use_skimage_features: bool) -> None:
+        self.logger.info("on_model_fit")
+        model = task.result()
+        self.model = model
         if live_prediction and self.model:
             # Update prediction_data
             if use_skimage_features:
                 prediction_features = np.array(self.feature_data_skimage)
             else:
                 prediction_features = np.array(self.feature_data_tomotwin)
-            # Add 1 becasue of the background label adjustment for the model
-            prediction = self.predict(
-                self.model, prediction_features, model_type
-            )
-            self.logger.info(
-                f"prediction {prediction.shape} prediction layer {self.get_prediction_layer().data.shape} prediction {np.transpose(prediction).shape} features {prediction_features.shape}"
-            )
+            
+            self.widget.status.setText("Predicting labels ...")
+            self.predict_task = self.executor.submit(self.predict, prediction_features)
+            self.predict_task.add_done_callback(self.on_prediction)
+
+    @ensure_main_thread
+    def on_prediction(self, task: Future[np.ndarray]) -> None:
+        prediction = task.result()
+        layer = self.get_prediction_layer()
+        self.logger.info(
+            f"prediction {prediction.shape} prediction layer {layer.data.shape} prediction {np.transpose(prediction).shape}"# features {prediction_features.shape}"
+        )
+        layer.data = np.transpose(prediction)
+        # Ensure the prediction layer visual is updated
+        # layer.refresh()
```

Comment:
The explicit `refresh` shouldn't be needed as the mutation of `layer.data` should be enough. It was for me, but maybe that's not true on all napari versions (I'm using v0.4.19).